### PR TITLE
perf: bypass duplexify

### PR DESCRIPTION
**Pull request by ronag**
Bypasses duplexify and pipes encode, socket, decode directly without any intermediate proxy.

### CODE REVIEW

**Comment by ronag** [comment]
Would be nice if multileveldown provided a less hacky way to do this...

File: index.js [hunk 1]
```diff
@@ -120,3 +119,14 @@ module.exports = function (dir, opts = {}) {
...
 }
+
+function createRpcStream (client, socket) {
+  socket.setWritable = function (stream) {
+    pump(socket, stream, () => {})
+  }
+  socket.setReadable = function (stream) {
+    pump(stream, socket, () => {})
+  }
+  const multileveldown = client._db.db // Grab multileveldown instance.
```
Comment (ronag) [inline index.js:130]:
```suggestion
  const multileveldown = client._db.db // HACK: Grab multileveldown instance.
```

**Comment by vweevers** [comment]
The goal here is to skip this line, if I understand it correctly? https://github.com/Level/multileveldown/blob/52f4861fc37500bfed167114f007bfce3104fc77/leveldown.js#L79.

**Comment by ronag** [comment]
> The goal here is to skip this line, if I understand it correctly? https://github.com/Level/multileveldown/blob/52f4861fc37500bfed167114f007bfce3104fc77/leveldown.js#L79.

Yep and avoid the overhead/complexity of an intermediate stream.

**Comment by vweevers** [comment]
The less hacky way would be to support a second argument here, defaulting to null:

https://github.com/Level/multileveldown/blob/52f4861fc37500bfed167114f007bfce3104fc77/client.js#L21-L23

**Comment by ronag** [comment]
> The less hacky way would be to support a second argument here, defaulting to null:

Yep. But that would require updating multileveldown.

**Comment by vweevers** [comment]
I worry this will have a side effects, because:

- `duplexify` handles a bunch of edge cases
- the `close` event on a net socket has a `hadError` argument that we may need to ignore [here](https://github.com/Level/multileveldown/blob/52f4861fc37500bfed167114f007bfce3104fc77/leveldown.js#L232)

**Comment by ronag** [comment]
> I worry this will have a side effects, because:
> 
> * `duplexify` handles a bunch of edge cases

I'm more worried about having duplexify than not having it.

> * the `close` event on a net socket has a `hadError` argument that we may need to ignore [here](https://github.com/Level/multileveldown/blob/52f4861fc37500bfed167114f007bfce3104fc77/leveldown.js#L232)

Yes.

**Comment by vweevers** [comment]
This makes no difference in my benchmarks, at least not after removing much bigger bottlenecks. In `many-level` (the WIP replacement for `level-party`) I doubled the throughput by batching reads, among other things.

**Comment by vweevers** [comment]
This makes no difference in my benchmarks, at least not after removing much bigger bottlenecks. In `many-level` (the WIP replacement for `level-party`) I doubled the throughput by batching reads, among other things.

**Comment by ronag** [comment]
Can I follow many-level somewhere?

**Comment by vweevers** [comment]
I'll post something in https://github.com/Level/abstract-level/issues/14 once I've pushed the repo

**Pull request closed without merging**